src/NuGet.Core/NuGet.Commands/RestoreCommand/ContentFiles/ContentFileUtils.cs

```diff
@@ -129,17 +129,16 @@ internal static class ContentFileUtils
                         var relativePath = file.Substring(rootFolderPathLength, file.Length - rootFolderPathLength);
 
                         // Check if the nuspec group include/exclude patterns apply to the file
-                        var virtualDirectory = new VirtualFileProvider(new List<string>() { relativePath });
                         var globbingDirectory = new FileProviderGlobbingDirectory(
-                            virtualDirectory,
+                            fileProvider: new SingleFileProvider(relativePath),
                             fileInfo: rootDirectory,
                             parent: null);
 
                         // Currently Matcher only returns the file name not the full path, each file must be
                         // check individually.
                         var matchResults = matcher.Execute(globbingDirectory);
 
-                        if (matchResults.Files.Any())
+                        if (matchResults.HasMatches)
```

Comment:
In case you haven't already figured out what this code is doing, is it's matching `contentFiles/*` files in the zip to `<contentFiles>` entries in the package's `nuspec`.  I'm guessing the reason this code is using the file globbing library rather than doing exact string matches, is because it's possible for the nuspec to use a `**/*` (or other) globs.

Therefore, I belive it's possible for one glob in the nuspec to match multiple files in the zip's `contentFiles/` directory. Using NuGet.Insights, I can see there are some packages that indeed use `*` in the nuspec's contentFiles definitions.  Are there any tests to ensure that this still works?

From a quick read, it looks like the `foreach` loop on line 120 is enumerating the files in the zip and checking them one by one against the nuspec (line 105's foreach). So, I guess this PR is safe, since the code indeed checks one file at a time.

I wonder if it would be more efficient to create a single instance of an optimized virtual file provider outside line 105's foreach loop, and that way we can eliminate line 120's foreach loop, since the file globbing library can do its work?

A more impactful perf improvement would be to stop using the file globbing library when the nuspec entry does not contain any `*` or `?` characters.

Reply:
That's a good observation and would be a nice perf improvement. Should we file a new issue for that, or would you want to see that in this PR too?

Reply:
new issue is fine. I just wanted to raise awareness since you seem to be spending a lot of time of all these nuget perf improvements. I didn't know if you'd be motivated to do it right away, or leave it as something for the future.

I just want to make sure that nuspecs that use globbing don't break though. I don't have confidence in NuGet's existing unit tests, and Kartheek's comment above is an example of why: https://github.com/NuGet/NuGet.Client/pull/5231#discussion_r1237802635

Reply:
Can we get a new issue with that idea?

Reply:
* https://github.com/NuGet/Home/issues/12695
* https://github.com/NuGet/Home/issues/12696